Add typing to dbcore types modules

Serene-Arc opened the pull request:
## Description

This is part of  #4484, adding typings to the `dbcore` submodule. It's very much still a work in progress. I've made a couple of `# TYPING ERROR` comments where there are obvious things that don't match up with what the typings would seem to suggest. MyPy gives a *lot* more errors than my IDE though so there's clearly some work to be done.

Serene-Arc commented:
So given that I'd like to try converting beets into using `pathlib`, I thought I'd try typing the path functions that beets uses. Honestly, I don't think it's possible though. Due to Python's ambiguous treatment of bytes and strings, there is a lot of code that doesn't distinguish, or only distinguishes some of the time, despite what it apparently says. 

For pretty much every function, the best that I can do is say it's either a string or a bytestring and leave it at that because anything more is just not possible. If this were statically typed then I honestly think it wouldn't even compile but because of Python's duck-typing system, I would imagine that it just works because it converts when it needs to and doesn't when it doesn't.

I'm happy to try opening another PR with the `pathlib` conversion and typing that as I go. That'll solve both problems at once and given that there's a large test base, it *should* be fine.

wisp3rwind commented:
> So given that I'd like to try converting beets into using `pathlib`, I thought I'd try typing the path functions that beets uses. Honestly, I don't think it's possible though. Due to Python's ambiguous treatment of bytes and strings, there is a lot of code that doesn't distinguish, or only distinguishes some of the time, despite what it apparently says.

In fact, beets has a convention of using ` bytestring`  for all internal paths, and only convert to anything else (such as `str`) at the boundary to external APIs. As such, I think any deviation from this should in fact be considered a bug. But, I'm also pretty sure that there are in fact such bugs still around. See for example https://github.com/beetbox/beets/pull/3690 which unfortunately I never got around to finishing up.

> I'm happy to try opening another PR with the `pathlib` conversion and typing that as I go. That'll solve both problems at once and given that there's a large test base, it _should_ be fine.

Skipping paths for now sounds fine! I'm pretty sure that our test suite has blind spots in this regard, however. It'll hopefully prevent most regressions, however.

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -71,10 +84,10 @@ def __getitem__(self, key):
             raise KeyError(key)
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable:
```
Couldn't this be made more specific as `Iterable[str]`?

Serene-Arc replied:
Fixed in e57d66e2

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -146,12 +159,12 @@ def __delitem__(self, key):
             del self.data[key]
 
-    def keys(self):
+    def keys(self) -> List:
```
`List[str]`?

Serene-Arc replied:
Fixed in e57d66e2

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -183,12 +196,12 @@ def get(self, key, default=None):
         return key in self.keys()
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable:
```
` Iterable[str]`?

Serene-Arc replied:
Fixed in e57d66e2

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -167,7 +180,7 @@ def update(self, values):
             self[key] = value
 
-    def items(self):
+    def items(self) -> Generator[Tuple, None, None]:
```
According to https://docs.python.org/3/library/typing.html#typing.Generator, this could also be written more concisely as `-> Iterable[Tuple[str, Any]]`.

Serene-Arc replied:
IMO it's better to type a generator as a generator, since it can have different uses to a more general iterable. IT's rare but people do call specific generator functions such as `next()`.

wisp3rwind replied:
Sure, in fact I'm not at all certain that I got these suggestions for `Iterator`, `Iterable`,  and `Generator` right. But the example that you mention seems to be about the distinction between the former two (`Iterator`  has `__next__` , `Iterable` only has `__iter__`). So, `from my understanding, Iterator`  would in fact be correct here, since not full generality of generators is allowed. `Generator[Tuple[..], None, None]`, seems to be synonymous, and I don't feel strongly about using either of the two expresssions.

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -267,14 +280,14 @@ class Model:
         raise NotImplementedError()
 
-    def _template_funcs(self):
+    def _template_funcs(self) -> NoReturn:
```
Is it really correct to type this as `NoReturn`, when it is expected to be overridden with a function `-> Mapping[str, Callable[[str], str]]`   ? (not sure that I got the callable right, should check actual users of `_template_funcs`)

Serene-Arc replied:
I honestly don't know the conventions here. It might be a matter of preference and style whether to mark functions that should be overloaded. Perhaps it will help for IDEs/linters? It's a minor thing, but the typing does exist and this is its explicit purpose in the Python docs.

wisp3rwind replied:
Yes, I don't know either what the "recommended" approach is. However, in the present case of an abstract method it seems more natural to me to declare the interface that should be implemented, rather than that of the stub method in the super class. I couldn't really find many examples for this; the mypy docs do something similar at https://mypy.readthedocs.io/en/stable/error_code_list.html?highlight=notimplementederror#check-that-call-to-an-abstract-method-via-super-is-valid-safe-super

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -455,19 +473,19 @@ def update(self, values):
             self[key] = value
 
-    def items(self):
+    def items(self) -> Generator:
```
`-> Iterator[Tuple[str, Any]]`

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -455,19 +473,19 @@ def update(self, values):
         return key in self.keys(computed=True)
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable:
```
` Iterable[str]`?

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -628,28 +654,37 @@ def evaluate_template(self, template, for_path=False):
 
     @classmethod
-    def _parse(cls, key, string):
+    def _parse(cls, key, string: str) -> types.Type:
```
I think `parse` actually return the field type ( `int`, `bool`, etc), so this should probably be `-> Any`.

wisp3rwind commented on beets/dbcore/db.py:
```diff
@@ -701,7 +743,7 @@ def __init__(self, model_class, rows, db, flex_rows,
         self._objects = []
 
-    def _get_objects(self):
+    def _get_objects(self) -> Generator[Model, None, None]:
```
`Iterable[Model]` in order to match `__iter__` below.

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -355,7 +370,7 @@ def _fsencoding():
 
 
-def bytestring_path(path):
+def bytestring_path(path: Bytestring) -> bytes:
```
Should probably take `Union[Bytestring, str]`)? Or maybe even `Union[bytes, str]`? I don't think we expect any other `Bytestring` subtype here.

Serene-Arc replied:
This is my defined type so it already is the union. I'll change the name to something less confusing in a coming commit.

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -376,10 +391,10 @@ def bytestring_path(path):
 
 
-def displayable_path(path, separator='; '):
+def displayable_path(path, separator: str = '; ') -> str:
```
Should be `path: bytes`,  I think. If it's not, that would be a bug in the caller.

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -398,7 +413,7 @@ def displayable_path(path, separator='; '):
 
 
-def syspath(path, prefix=True):
+def syspath(path: Bytestring, prefix: bool = True) -> Bytestring:
```
```suggestion
def syspath(path: bytes, prefix: bool = True) -> Union[bytes, str]:
```

Output type is platform-dependent, unfortunately. Maybe, there's a way to encode this is in the types?

Serene-Arc replied:
No, there isn't. That isn't how the typing works. The way to do that would be different functions that are typed differently. 

I defined `Bytestring` as equivalent to the union of strings and bytes. Any reason why you changed it? I think the return might be a string no matter what since the conversion happens but that has knock on effects with the rest of the code.

Also `path` can clearly be a string, not bytes, because there's an `isinstance` check in that function that sees whether it's in string or byte form.

wisp3rwind replied:
Oh, what happened here is that I confused the custom `Bytestring` with [`typing.ByteString`](https://docs.python.org/3/library/typing.html#typing.ByteString). Seems useful to have an alias here, but we may want to come up with a dfferent name.

> I think the return might be a string no matter what since the conversion happens but that has knock on effects with the rest of the code.

No, the conversion does not always happen, only on Windows.

> Also path can clearly be a string, not bytes, because there's an isinstance check in that function that sees whether it's in string or byte form.

You're right, that is indeed the case. From my understanding, correct usage of that function shouldn't pass `str`,  but it probably makes sense to keep that check as a safeguard for now.

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -432,14 +448,14 @@ def syspath(path, prefix=True):
 
 
-def remove(path, soft=True):
+def remove(path: AnyStr, soft: bool = True):
```
`path: bytes` (same for the functions below)

Serene-Arc replied:
Fixed in 5abe5d54

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -636,7 +657,10 @@ def unique_path(path):
 
 
-def sanitize_path(path, replacements=None):
+def sanitize_path(
+        path: Bytestring,
+        replacements: Optional[Sequence[Sequence[Pattern | str]]] = None,
+) -> str | bytes:
```
Shouldn't this be `str -> str` ?

Serene-Arc replied:
If that's the case then there are flow-in effects. The `path` variable goes through `_legalize_stage()` and `legalize_path()` without being changed, so if `sanitize_path` only accepts strings, then so must those functions to keep it consistent.

I've made these changes in 9453a513

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -762,12 +798,12 @@ def py3_path(path):
 
 
-def as_string(value):
+def as_string(value: Optional[memoryview | bytes]) -> str:
```
Need to check the call sites, but if I remember correctly this can take `Any`, not just `bytes`.

Serene-Arc replied:
Quite possibly, these are just the usages that occur in this file. Once it's expanded to the entirety of the project, many typings will likely have to be broadened.

Serene-Arc replied:
Oops wrong one

Serene-Arc replied:
Fixed in 4d1dec3c

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -1056,7 +1092,7 @@ def asciify_path(path, sep_replace):
 
 
-def par_map(transform, items):
+def par_map(transform: Callable, items: Sequence):
```
```suggestion
def par_map(transform: Callable, items: Iterable):
```

wisp3rwind reviewed (commented):
Had a quick look through this (without actually running mypy, though); here are a few comments on where I think the types can be made more precise or are not quite right.

Serene-Arc commented:
This all makes me think `pathlib` should be implemented ASAP lol

Serene-Arc commented:
@wisp3rwind If I were to go about converting to pathlib in another PR, would that be fine vis-a-vis this PR? It would make all of the commits for the util file obsolete but there wouldn't be any confusion and it would be much simpler to type.

wisp3rwind commented:
> @wisp3rwind If I were to go about converting to pathlib in another PR, would that be fine vis-a-vis this PR? It would make all of the commits for the util file obsolete but there wouldn't be any confusion and it would be much simpler to type.

I'm not quite sure that I follow your question here: pathlib and the current PR seem to be mostly independent projects, and I don't think these initial typings need to be completely perfect before we merge this (seems to be unrealistic, since we can't really run mypy on the full codebase yet). Converting to pathlib seems like a huge project however, and I suspect that having (partial) type annotations might actually help with the conversion (so, I'd keep the `util` part of this PR). Also, as discussed at https://github.com/beetbox/beets/issues/1409 and https://github.com/beetbox/beets/pull/3690#issuecomment-664074016, its not completely obvious that pathlib can get us rid of all manual conversions (it might, but nobody has really thought about this yet).

Serene-Arc commented:
I intended to try my hand at the pathlib conversion soon but I mentioned a while back it might be good to try this typing project beforehand to make sure that there were fewer issues. That might still be the case but it seems to have less of an impact.

I've used pathlib extensively and in general I don't think conversions are required. The `Path` object accepts bytes or strings and then can be passed to all of the other standard library functions. Windows and Linux can be treated more or less the same in my other projects once the strings have been sanitised. IIRC the only problems that we had were with malformed Unicode bytes.

Serene-Arc commented:
I've addressed all of the issues that you've raised but I think there's a lot of work to be done yet. Do you have a direction you'd like me to take? Way I see it, there are two main paths:

1. Start using mypy to firm down the typings in the modules that I've already started
2. Expand the typings project to other modules, keeping the rough ones and using mypy at the end on the project as a whole to resolve conflicting typings

wisp3rwind commented:
Hey, sorry for the late reply; the time that I can invest into beets its fairly limited these days!

> I intended to try my hand at the pathlib conversion soon but I mentioned a while back it might be good to try this typing project beforehand to make sure that there were fewer issues. That might still be the case but it seems to have less of an impact.
> 
> I've used pathlib extensively and in general I don't think conversions are required. The `Path` object accepts bytes or strings and then can be passed to all of the other standard library functions. Windows and Linux can be treated more or less the same in my other projects once the strings have been sanitised. IIRC the only problems that we had were with malformed Unicode bytes.

One point of friction might be the connection of pathlib to our database model (which always stored bytestrings). But you're probably better qualified to talk about this than me; my own usage of pathlib has been fairly limited and mostly in ASCII-only scenarios. I think the only real constraint should be backwards-compatibility (or, conscious decisions to break it with appropriate warnings to users).


> I've addressed all of the issues that you've raised but I think there's a lot of work to be done yet. Do you have a direction you'd like me to take? Way I see it, there are two main paths:
> 
> 1. Start using mypy to firm down the typings in the modules that I've already started
> 
> 2. Expand the typings project to other modules, keeping the rough ones and using mypy at the end on the project as a whole to resolve conflicting typings

I guess I'd like to see mypy rather sooner than later (I also find its errors useful when adding missing types); but in principle, whatever motivates you!

From the reviewers perspective, I think it would be very helpful to split things into manageable PRs, even that implies that they might not be perfect initially (which is harmless, after all).
In particular, could you move the new commits for the autotag module to a separate PR (relative to this one)? Then, I'd give the dbcore types another quick glance and merge it. Similarly, mypy could be added to the CI in a PR of its own.

Serene-Arc commented:
Sure I can do that. Mypy is more strict so there will likely have to be changes. Also, there are additional changes to the `dbcore` typing that are becoming clear as I move through modules that reference them, so hopefully that won't cause issues with the branch splitting.

Serene-Arc commented:
This commit here should be pretty good for merging.

wisp3rwind commented:
> Sure I can do that. Mypy is more strict so there will likely have to be changes. Also, there are additional changes to the `dbcore` typing that are becoming clear as I move through modules that reference them, so hopefully that won't cause issues with the branch splitting.

Thanks!

I guess we don't need to enforce mypy compliance (yet), but it would be nice to have it available via tox (and maybe even in CI, but without actually failing the entire build, similar to how we test python alphas?). I might actually have a look at that myself.

wisp3rwind commented:
> > Sure I can do that. Mypy is more strict so there will likely have to be changes. Also, there are additional changes to the `dbcore` typing that are becoming clear as I move through modules that reference them, so hopefully that won't cause issues with the branch splitting.
> 
> Thanks!
> 
> I guess we don't need to enforce mypy compliance (yet), but it would be nice to have it available via tox (and maybe even in CI, but without actually failing the entire build, similar to how we test python alphas?). I might actually have a look at that myself.

About issues with branch splitting: I'm not sure I get what you mean, but wouldn't you avoid any problem by rebasing the autotag one on top of this one?

wisp3rwind commented:
When this is ready for a final review, can you mark the PR as such?

@sampsyo (and anyone else!): Any concerns on going ahead with typings in beets?

---

One thought that I had when thinking about paths in beets: It might be a good idea to have newtypes for bytes and str here to encode the semantics more clearly, i.e.

```python
InternalPath = NewType("InternalPath", bytes)
SysPath = NewType("SysPath", Union[bytes, str])  # <- or, maybe, define this per-platform?

def syspath(path: InternalPath) -> SysPath:
    ...

def displayable_path(path: InternalPath) -> str:
    ...

# etc
```

(Doesn't have to happen in this PR already, but I'd be happy about any opinions!)

Serene-Arc commented:
@wisp3rwind I suppose it wouldn't be an issue for the branch splitting if I base it off this one but generally I like to hold to the principle that each PR is a self-contained change to the development or master branch. Basing PRs off of other PRs that have yet to be merged sounds like a recipe for chaos, but if it's allowed then sure, I can do that.

I think that it's ready for final review now, unless I need to make it mypy passable right now?

The internal path type alias can be done, but like I said above, there's no way to type in a platform dependent manner. The way to do that is to have separate functions, there's no way in Python to define the same function parameter to be bytes on windows and strings everywhere else.

wisp3rwind commented on beets/dbcore/query.py:
```diff
@@ -163,14 +169,18 @@ class StringFieldQuery(FieldQuery):
 
     @classmethod
-    def value_match(cls, pattern, value):
+    def value_match(cls: Type['StringFieldQuery'], pattern: str, value: Any):
```
Let's just drop the type for `cls` here?  I suppose type checkers don't need it, and if we really need to type `cls`, I think we should wait until we drop support for older Python which doesn't have the `Self` type yet. (Same for `cls` annotations below.)

wisp3rwind reviewed (changes requested):
Hi @Serene-Arc , I've finally looked through the entire changeset, and left a bunch of inline comments. I think I'd be good to address at least the ones for `util/__init__.py` before merging this.

For `dbcore` my impression is that, after all, thi is not the easiest component of beets to type: There are a lot of abstract base classes, things that seem to require what Rust calls "associated types", and I guess we will quite a few iterations to really nail down the typing there (and potentially, first fix the code to remove some nasty hacks). Thus, I would merge this without going through all the comments that I left.

wisp3rwind commented:
> @wisp3rwind I suppose it wouldn't be an issue for the branch splitting if I base it off this one but generally I like to hold to the principle that each PR is a self-contained change to the development or master branch. Basing PRs off of other PRs that have yet to be merged sounds like a recipe for chaos, but if it's allowed then sure, I can do that.

Fair enough. However we do it exactly, what I would really like to achieve is to keep these PRs relatively small. This one is already rather large, and reviewing it entirely literally takes hours, and I already looked at some parts quite superficially.

> I think that it's ready for final review now, unless I need to make it mypy passable right now?

No, let's not bother too much with making mypy happy for now.

> The internal path type alias can be done, but like I said above, there's no way to type in a platform dependent manner. The way to do that is to have separate functions, there's no way in Python to define the same function parameter to be bytes on windows and strings everywhere else.

Yeah, I'm not sure about how to do this either, and it's certainly something for a separate PR. What I had in mind is something along the lines of
```python
import sys
if sys.platform == "win32":
    SysPath = NewType("SysPath", str)
else:
    SysPath = NewType("SysPath", bytes)
```
but the question is of course whether type checkers can deal with that. According to https://mypy.readthedocs.io/en/stable/common_issues.html#version-and-platform-checks, it might actually work.

Serene-Arc commented:
The `dbcore` modules were recommended to me as the 'base' of the system so the typings would be easier to propogate out but in retrospect, not the easiest place to start.

Serene-Arc commented on beets/util/__init__.py:
```diff
@@ -136,7 +142,7 @@ class MoveOperation(Enum):
     REFLINK_AUTO = 5
 
 
-def normpath(path):
```
Not that I know of, changed it.

Serene-Arc commented on beets/util/__init__.py:
```diff
@@ -145,7 +151,7 @@ def normpath(path):
     return bytestring_path(path)
 
 
-def ancestry(path):
```
`AnyStr` means that it is bytes or a string but in a way that is consistent through the function e.g. if there are multiple instances of variables of type `AnyStr` they are all either bytes or all strings.

For a single instance it is essentially a union when viewed as a single function but mypy might propagate the typing restrictions down the stack, I don't know enough about it to be sure.

wisp3rwind replied:
Exactly; what I meant to suggest is that this should be `def ancestry(path: AnyStr) -> List[AnyStr]:`

Serene-Arc commented on beets/util/__init__.py:
```diff
@@ -226,14 +237,14 @@ def sorted_walk(path, ignore=(), ignore_hidden=False, logger=None):
 
 
-def mkdirall(path):
+def mkdirall(path: str):
```
This function calls `ancestry()` which has a `str` in the function description example. If that is wrong, then it should be changed to bytes as it's called directly with the path parameter used here.

Serene-Arc commented on beets/util/__init__.py:
```diff
@@ -686,7 +716,7 @@ def _legalize_stage(path, replacements, length, extension, fragment):
     # Encode for the filesystem.
     if not fragment:
-        path = bytestring_path(path)
+        path = bytestring_path(path)  # type: ignore
```
Hopefully once I get the pathlib PR going this won't be an issue. This was mostly for mypy since it doesn't like when you change the type of named variables, it always throws an error.

wisp3rwind replied:
mypy complains for good reason, changing the type in this way really seems like a recipe for disaster. I'm not sure that pathlib magically makes this disappear, I suspect that the callers of this function might still need a different type depending on whether dealing with path fragments or an entire path.

Serene-Arc commented on beets/dbcore/types.py:
```diff
@@ -35,7 +36,7 @@ class Type:
     """
 
-    model_type = str
+    model_type: Callable[[Any], str] = str
```
Right now this is saying that it takes in any parameters and outputs a str, and is defaulting to the builtin function `str()`. Not the clearest syntax.

I don't know if there's a way to do that with a TypeVar since you would have to dynamically bind it to the builtin cast function, seems to me like that would require another if-else block to do easily. 

Plus, for this function, we know that it always returns a string since it's bound to the string cast function, so a TypeVar would actually be incorrect as it's always going to be a string.

wisp3rwind replied:
Oh, ok, there seem to be two approaches how to model the types here:

- what you did, understand `model_type` as the function which parses `Any` -> `<the target type>`, which works since classes `int`, `float`, etc are also callable objects which return an instance of that class
- what I had in mind: `model_type` is the class that values of the given type should have, which also happens to be a callable

However, I'm pretty sure that the output type is not actually `str`, but at this point you cannot be more precise than `Any`. For example, consider when `model_type = int`.

The best way to express all of this might be using generic, cf. https://stackoverflow.com/questions/74885777/how-to-implement-associated-types-in-python-mypy-or-what-to-do-when-wanting-su

Let's not do that now I'd say, but go for the `Callable` approach.

Serene-Arc replied:
So leave as is?

Serene-Arc commented on beets/dbcore/queryparse.py:
```diff
@@ -158,7 +172,13 @@ def construct_query_part(model_cls, prefixes, query_part):
 
 
-def query_from_strings(query_cls, model_cls, prefixes, query_parts):
+# TYPING ERROR
```
I think that it doesn't fit the `__init__` parameters of every child of the `Query` class. It's submitting a list of queries to the `__init__` of the specified class but a lot of classes require more than one parameter. My IDE is highlighting it as an issue.

I should do more detailed comments.

wisp3rwind replied:
Hmm, I'm not quite following here. I'll mark this as resolved for now, and will have a look when this is merged.

Serene-Arc commented on beets/dbcore/queryparse.py:
```diff
@@ -34,8 +38,14 @@
 
-def parse_query_part(part, query_classes={}, prefixes={},
-                     default_class=query.SubstringQuery):
+# TYPING ERROR
+# not sure what the query type should be
```
Apparently no I can't because I have no idea what I was referring to. Seems to have been resolved by other changes or is somewhere in the usage. I'll remove the comments.

Serene-Arc commented on beets/dbcore/types.py:
```diff
@@ -125,7 +131,7 @@ class Integer(Type):
     model_type = int
 
-    def normalize(self, value):
+    def normalize(self, value: str) -> Union[int, str]:
```
We could define a type alias per class?

wisp3rwind replied:
Seems like the solution is to make `Type` generic over `model_type` and `null_type` cf. https://stackoverflow.com/questions/74885777/how-to-implement-associated-types-in-python-mypy-or-what-to-do-when-wanting-su

How much time would you still like to invest into this project? If you like, go ahead, otherwise, I'd also offer to merge this quickly, and help with improving it myself afterwards.

Serene-Arc replied:
I'm happy to invest more time into this but maybe in another PR? You seem to have a good idea of what you'd like the system to be for these classes' typing, and it would also mean I can work on the pathlib PR more easily.

wisp3rwind commented:
> The `dbcore` modules were recommended to me as the 'base' of the system so the typings would be easier to propogate out but in retrospect, not the easiest place to start.

Sorry for that! Thanks a lot for sticking around anyway!

wisp3rwind commented on beets/util/__init__.py:
```diff
@@ -376,10 +391,10 @@ def bytestring_path(path):
 
 
-def displayable_path(path, separator='; '):
+def displayable_path(path: bytes, separator: str = '; ') -> str:
```
I think you accidentally added the comment to `bytestring_path`?

wisp3rwind reviewed (commented):
I left a few more small comments, what do you think?

Serene-Arc commented:
I've left comments on them but more broadly, if you'd like to merge what there is now, I'd be fine with that. As I expand the typings across modules, there's going to be at least some changes and merging nowish would make that easier, plus allow a bit more collaboration on the different parts.

Sorry that this turned into a rather mammoth and wide-reaching PR, there's a lot to type.

JOJ0 commented:
Without having an idea about the details and what's left but it seems like that you both agreed that this could be merged now and the "typing project" will be tackled further in subsequent PR's. Should we just merge it and if you think it's required leave a note in form of an issue of what's left to be "typed"?

Serene-Arc commented:
@JOJ0 yes I think that this can be merged. There isn't likely to be any actual issues since it's all typings and mypy hasn't been integrated into the checks yet.

Serene-Arc commented:
I need to resolve the conflicts but my brain isn't working right now and I keep messing up, so I'll get to it soon.

JOJ0 commented:
LOL yeah I know that issue. Resolving conflicts sometimes requires more cpu-power than it deserves!

wisp3rwind commented:
Thanks for wrapping this up :tada:

Serene-Arc commented:
Finally got the tests working and the conflicts resolved!
